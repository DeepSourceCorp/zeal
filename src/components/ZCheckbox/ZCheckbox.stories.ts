import '../../assets/css/tailwind.css'
import '../../assets/css/typography.css'
import '../../assets/css/layout.css'
import ZCheckbox from './ZCheckbox.vue'

export default {
  title: 'Checkbox',
  component: ZCheckbox,
  excludeStories: /.*Data$/
}

export const DefaultCheckbox = () => ({
  components: { ZCheckbox },
  data() {
    return {
      selectedValue: false
    }
  },
  template: `<z-checkbox label="I love Pastries" value="pastry" v-model="selectedValue"/>`
})

export const DisabledCheckbox = () => ({
  components: { ZCheckbox },
  data() {
    return {
      selectedValue: false
    }
  },
  template: `<z-checkbox label="I love Pastries" value="pastry" v-model="selectedValue" disabled/>`
})

export const SelectedCheckboxWithDisabledProperty = () => ({
  components: { ZCheckbox },
  data() {
    return {
      selectedValue: true
    }
  },
  template: `<z-checkbox label="I love Pastries" value="pastry" v-model="selectedValue" disabled/>`
})

export const SelectedCheckbox = () => ({
  components: { ZCheckbox },
  data() {
    return {
      selectedValue: true
    }
  },
  template: `<z-checkbox label="I love Pastries" value="pastry" v-model="selectedValue"/>`
})

export const CheckboxGroup = () => ({
  components: { ZCheckbox },
  data() {
    return {
      selectedValues: ['pizza', 'coding']
    }
  },
  template: `<div class="container">
    <z-checkbox label="Pizza" value="pizza" v-model="selectedValues"/>
    <z-checkbox label="Coding" value="coding" v-model="selectedValues"/>
    <z-checkbox label="Walk by myself to get water" value="walk" v-model="selectedValues"/>
  </div>`
})
