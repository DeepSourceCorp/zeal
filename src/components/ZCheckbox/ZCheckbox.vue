<template>
  <label class="z-input z-input__checkbox block relative pl-6 mb-1 text-base" 
        :class="[
          `${disabled && `cursor-not-allowed`}`,
          { 'is-disabled': this.disabled !== false }
        ]">
    <input
      type="checkbox"
      :checked="isChecked"
      :value="value"
      @change="updateInput"
      :disabled="this.disabled"
      class="absolute opacity-0 h-0 w-0"
    />
    <span class="z-input__checkbox--checkmark grid place-items-center absolute left-0 h-5 w-5 rounded-sm border border-slate border-solid bg-transparent cursor-pointer top-50"
          :class="[
            `${disabled && `cursor-not-allowed`}`
          ]"></span>
    <span class="z-input__checkbox--text ml-2"
          :class="[
            `${disabled && `text-slate`}`
          ]">
      {{ label }}
    </span>
  </label>
</template>

<script>
export default {
  name: 'ZCheckbox',
  model: {
    prop: 'modelValue',
    event: 'change'
  },
  props: {
    label: {
      default: '',
      type: String
    },
    value: {
      default: '',
      type: String
    },
    trueValue: {
      default: true,
      type: Boolean
    },
    falseValue: {
      default: false,
      type: Boolean
    },
    disabled: {
      default: false,
      type: Boolean
    },
    modelValue: {
      default: ''
    }
  },
  computed: {
    isChecked() {
      if (this.modelValue instanceof Array) {
        return this.modelValue.includes(this.value)
      }
      return this.modelValue === this.trueValue
    }
  },
  methods: {
    updateInput(event) {
      let isChecked = event.target.checked

      if (this.modelValue instanceof Array) {
        let newValue = [...this.modelValue]

        if (isChecked) {
          newValue.push(this.value)
        } else {
          newValue.splice(newValue.indexOf(this.value), 1)
        }
        this.$emit('change', newValue)
      } else {
        this.$emit('change', isChecked ? this.trueValue : this.falseValue)
      }
    }
  }
}
</script>

<style lang="css" scoped>
/* Create a custom checkbox */
.z-input__checkbox--checkmark {
  transform: translateY(-50%);
}

/* On mouse-over, add a grey background color */
.z-input__checkbox:hover input ~ .z-input__checkbox--checkmark {
  @apply border border-solid border-juniper;
}

/* When the checkbox is checked, add a blue background */
.z-input__checkbox input:checked ~ .z-input__checkbox--checkmark {
  @apply border border-solid border-juniper bg-juniper;
}

/* Create the checkmark/indicator (hidden when not checked) */
.z-input__checkbox--checkmark:after {
   
  @apply absolute hidden;
}

/* Show the checkmark when checked */
.z-input__checkbox input:checked ~ .z-input__checkbox--checkmark:after {
  @apply block;
}

/* Style the checkmark/indicator */
.z-input__checkbox .z-input__checkbox--checkmark:after {
  content: "\2713";
  @apply text-vanilla-100 text-center text-lg;
}

/* Styles for Disabled Proeprty */

.z-input__checkbox.is-disabled:hover .z-input__checkbox--checkmark {
  @apply border-slate;
}

.z-input__checkbox.is-disabled input:checked ~ .z-input__checkbox--checkmark {
  @apply border border-solid border-slate bg-slate;
}

.z-input__checkbox.is-disabled input:checked ~ .z-input__checkbox--checkmark:after {
  @apply text-ink-300;
}

</style>
