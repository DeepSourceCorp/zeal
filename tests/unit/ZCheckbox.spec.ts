import ZCheckbox from '../../src/components/ZCheckbox'
import Vue from 'vue'

import { createLocalVue, mount, Wrapper } from '@vue/test-utils'

describe('Checkbox Component', () => {
  let mountFn: (options?: object) => Wrapper<Vue>
  let localVue: typeof Vue

  beforeEach(() => {
    mountFn = (options = {}) => {
      return mount(ZCheckbox, {
        localVue,
        ...options
      })
    }
  })

  it('renders the checkbox component', () => {
    const wrapper = mountFn()

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.classes().includes('z-input__checkbox')).toBe(true)
  })

  it('renders a disabled checkbox component when disabled prop is passed', () => {
    const wrapper = mountFn({
      propsData: {
        disabled: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.classes().includes('is-disabled')).toBe(true)
  })

  it('renders a pre-selected checkbox when the value is passed as checked', () => {
    const wrapper = mountFn({
      propsData: {
        value: 'icecream',
        label: 'Icecreams'
      }
    })
    wrapper.setData({ selectedValue: true })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
