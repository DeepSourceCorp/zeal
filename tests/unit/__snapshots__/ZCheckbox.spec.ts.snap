// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component renders a disabled checkbox component when disabled prop is passed 1`] = `<label class="z-input z-input__checkbox is-disabled"><input type="checkbox" disabled="disabled" value=""> <span class="z-input__checkbox--checkmark"></span> <span class="z-input__checkbox--text"></span></label>`;

exports[`Checkbox Component renders a pre-selected checkbox when the value is passed as checked 1`] = `<label class="z-input z-input__checkbox"><input type="checkbox" value="icecream"> <span class="z-input__checkbox--checkmark"></span> <span class="z-input__checkbox--text">Icecreams</span></label>`;

exports[`Checkbox Component renders the checkbox component 1`] = `<label class="z-input z-input__checkbox"><input type="checkbox" value=""> <span class="z-input__checkbox--checkmark"></span> <span class="z-input__checkbox--text"></span></label>`;
